// sensor/sensor.js

module.exports = function(config,measurementDB,interval){
	
	this.buildExecString=function(config){
		var execString = '';
		if(config.superUserExec){
			execString += 'sudo ';
		}
		execString += 'python '
		execString += config.rundir
		execString += config.dir
		if(config.dummy){
			execString += config.dummyPrefix;
		}
		execString += config.path;
		if(config.interface=='spi'){
			execString += ' '
			execString += config.channel;
		}
		
		return execString;
	}
	this.sensorPath = this.buildExecString(config);
	this.intervalDuration = interval;
	this.sockets = [];
	this.measurementDB = measurementDB;
	this.name = config.name;
	this.eventEmitters = [];
	
	this.addEventEmitter = function(eventEmitter){
			
		this.eventEmitters.push(eventEmitter);
	}
	
	this.addSocket = function(socket){
		this.sockets.push(socket);
	}
	this.removeSocket = function(socket){
		var index = this.sockets.indexOf(socket);
		this.sockets.splice(index,1);
	}
	

	this.enable=function(){
		var slef = this;
		this.interval = setInterval(function(){slef.measure()},slef.intervalDuration);	
	}
	
	this.measure=function(){
		var my_command = this.sensorPath;
		console.log("execSwitch: "+my_command);
		var process = require('child_process');
		var self = this;
		var ls = process.exec(
					my_command,
					function(error, stdout, stderr){
						self.measure_cb(self, error, stdout, stderr);
					});
		ls.on('exit', function(code){ 
			console.log(self.name+' switch exit code: '+ code);
		});
	};
	
	this.measure_cb=function(sensor, error, stdout, stderr){
		if (error) {
			 console.log(error.stack);
			 console.log('Error code: '+error.code);
			 console.log('Signal received: '+error.signal);
		}else{
			var dateFormat = require('dateformat');
			var measurement = stdout.trim();
			var now = Date();
			for(var i=0; i < sensor.eventEmitters.length;i++){
				ee = sensor.eventEmitters[i];
				ee.emit(sensor.name,measurement)
			}
			
			for(var i=0; i<sensor.sockets.length; i++){
				sockets[i].emit(sensor.name, {temperature:temp});
			}
			sensor.measurementDB.insert({ date: dateFormat(now, "yyyy-mm-dd, HH:MM:ss"), value: measurement, sensor: sensor.name }, function (err) { });
			sensor.measurementDB.insert({ date: dateFormat(now, "yyyy-mm-dd, HH:MM:ss"), value: measurement, sensor: sensor.name }, function (err) { });
			console.log(sensor.name+' measurement: ' + stdout);
			
		};
	};
};
